packages/base/room.gts

```diff
@@ -0,0 +1,541 @@
+        newEvents.set(event.event_id, event);
+      }
+      return [...newEvents.values()];
+    },
+  });
+
+  @field roomId = contains(StringCard, {
+    computeVia: function (this: RoomCard) {
+      return this.events.length > 0 ? this.events[0].room_id : undefined;
+    },
+  });
+
+  @field name = contains(StringCard, {
+    computeVia: function (this: RoomCard) {
+      let roomState = roomStateCache.get(this);
+      if (!roomState) {
+        roomState = {} as RoomState;
+        roomStateCache.set(this, roomState);
+      }
+      let name = roomState.name;
+      // room name can change so we need to check new
+      // events for a room name even if we already have one
+      let events = this.newEvents
+        .filter((e) => e.type === 'm.room.name')
+        .sort((a, b) => a.origin_server_ts - b.origin_server_ts) as
+        | RoomNameEvent[];
+      if (events.length > 0) {
+        roomState.name = name ?? events.pop()!.content.name;
+      }
+      return roomState.name;
+    },
+  });
+
+  @field creator = contains(RoomMemberCard, {
+    computeVia: function (this: RoomCard) {
+      let roomState = roomStateCache.get(this);
+      if (!roomState) {
+        roomState = {} as RoomState;
+        roomStateCache.set(this, roomState);
+      }
+      let creator = roomState.creator;
+      if (creator) {
+        return creator;
+      }
+      let event = this.newEvents.find((e) => e.type === 'm.room.create') as
+        | RoomCreateEvent
+        | undefined;
+      if (event) {
+        roomState.creator = upsertRoomMember({
+          roomCard: this,
+          userId: event.sender,
+        });
+      }
+      return roomState.creator;
+    },
+  });
+
+  @field created = contains(DateTimeCard, {
+    computeVia: function (this: RoomCard) {
+      let roomState = roomStateCache.get(this);
+      if (!roomState) {
+        roomState = {} as RoomState;
+        roomStateCache.set(this, roomState);
+      }
+      let created = roomState.created;
+      if (created != null) {
+        return new Date(created);
+      }
+      let event = this.newEvents.find((e) => e.type === 'm.room.create') as
+        | RoomCreateEvent
+        | undefined;
+      if (event) {
+        roomState.created = event.origin_server_ts;
+      }
+      return roomState.created != null
+        ? new Date(roomState.created)
+        : roomState.created;
+    },
+  });
+
+  @field roomMembers = containsMany(RoomMemberCard, {
+    computeVia: function (this: RoomCard) {
+      let roomMembers = roomMemberCache.get(this);
+      if (!roomMembers) {
+        roomMembers = new Map();
+        roomMemberCache.set(this, roomMembers);
+      }
+
+      // The room member events are very much out of order chronologically,
+      // so using this.newEvents results in broken behavior as it is not
+      // uncommon to get an older invite event in a subsequent batch of events
+      // after the join event for a member has already been processed
```

Comment:
I should be able to solve this be adding a modified timestamp to the room member event and only considering events newer than the last event considered from the this.newEvents field